### PR TITLE
CSHARP-4831: Emit heartbeat started event before connection establishment in ServerMonitor

### PR DESCRIPTION
**Pull request by BorisDog**

### CODE REVIEW

File: tests/MongoDB.Driver.Tests/Specifications/server-discovery-and-monitoring/ServerDiscoveryAndMonitoringProseTests.cs [hunk 1]
```diff
@@ -64,21 +67,79 @@ public void Heartbeat_should_work_as_expected()
...
+                nameof(ServerHeartbeatStartedEvent),
+                HelloReceivedEvent,
+                ClientConnectedEvent,
+                nameof(ServerHeartbeatFailedEvent),
+            });
+
+            bool EnqueueEvent(string @event)
+            {
+                // Ignore RTT events
+                if (!Environment.StackTrace.Contains(nameof(RoundTripTimeMonitor)))
```
Comment (sanych-sun) [inline tests/MongoDB.Driver.Tests/Specifications/server-discovery-and-monitoring/ServerDiscoveryAndMonitoringProseTests.cs:131]: Minor: I would revert the condition:
```
if (Environment.StackTrace.Contains(nameof(RoundTripTimeMonitor)))
{
    return false;
}

 events.Enqueue(@event);
 return true;
```

**Review by sanych-sun — APPROVED** [review]
LGTM + minor comment

**Review by adelinowona — APPROVED** [review]
LGTM

File: tests/MongoDB.Driver.Tests/Specifications/server-discovery-and-monitoring/ServerDiscoveryAndMonitoringProseTests.cs (hunk 1 above)
Comment (BorisDog) [reply tests/MongoDB.Driver.Tests/Specifications/server-discovery-and-monitoring/ServerDiscoveryAndMonitoringProseTests.cs]: Thanks, done.

**Pull request merged**